### PR TITLE
Add goto Top button for Leads Page

### PR DESCRIPTION
**Pull request by subinasr**

### CODE REVIEW

**Review by samshara — APPROVED** [review]

File: src/views/Entries/QualityControl/index.tsx [hunk 1]
```diff
@@ -130,6 +132,33 @@ function QualityControl(props: Props) {
...
+                    const scrollTop = listContainerRef.current?.scrollTop ?? 0;
+                    setGotoTopButtonVisible(scrollTop > 0);
+                }, 200);
+            };
+            const reff = listContainerRef.current;
+            if (reff) {
+                reff.addEventListener('scroll', handleListScroll);
+            }
+            return () => {
+                if (reff) {
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx:154]: scrollTimeout should also be cleared on dismount

File: src/views/Entries/QualityControl/index.tsx [hunk 2]
```diff
@@ -148,6 +177,17 @@ function QualityControl(props: Props) {
...
+    const handleGotoTopButtonClick = useCallback(() => {
+        const c = listContainerRef.current;
+        if (c) {
+            c.scrollTo({
+                top: 0,
+                left: 0,
+                behavior: 'smooth',
+            });
+        }
+    }, [listContainerRef]);
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx]: This dependency is most probably not required

File: src/views/Entries/index.js [hunk 3]
```diff
@@ -525,6 +530,27 @@ export default class Entries extends React.PureComponent {
 
     handleHashChange = (view) => {
         this.setState({ view });
+
+        if (view === 'list') {
+            window.setTimeout(() => {
```
Comment (tnagorra) [inline src/views/Entries/index.js:546]: We should clear this timeout

File: src/views/Entries/index.js [hunk 4]
```diff
@@ -525,6 +530,27 @@ export default class Entries extends React.PureComponent {
...
+                        this.scrollTimeout = window.setTimeout(() => {
+                            this.setState({ gotoTopButtonVisible: e.target.scrollTop > 0 });
+                        }, 200);
+                    };
+
+                    c.addEventListener('scroll', this.handleListScroll);
+                }
+            }, 0);
+        } else {
+            window.removeEventListener('scroll', this.handleListScroll, true);
```
Comment (tnagorra) [inline src/views/Entries/index.js:551]: We should also clear timeout here

File: src/views/Entries/index.js [hunk 5]
```diff
@@ -558,9 +584,23 @@ export default class Entries extends React.PureComponent {
...
+        const handleGotoTopButtonClick = () => {
+            const c = this.listContainerRef.current;
+            if (c) {
+                c.scrollTo({
+                    top: 0,
+                    left: 0,
+                    behavior: 'smooth',
+                });
+            }
+        };
```
Comment (tnagorra) [inline src/views/Entries/index.js]: Why define this on render method!

File: src/views/Entries/index.js [hunk 6]
```diff
@@ -588,9 +628,16 @@ export default class Entries extends React.PureComponent {
                             onPageClick={this.handlePageClick}
                             showItemsPerPageChange={false}
                         />
+                        {this.state.gotoTopButtonVisible &&
```
Comment (tnagorra) [inline src/views/Entries/index.js]: Please de-structure  from this.state

File: src/views/Entries/index.js (hunk 4 above)
Comment (tnagorra) [inline src/views/Entries/index.js:553]: Please clearTimeout on this.scrollTimeout

File: src/views/Home/RecentActivity/index.tsx [hunk 7]
```diff
@@ -31,7 +31,6 @@ function RecentActivities() {
         url: 'server://projects/recent-activities/',
         method: 'GET',
         autoTrigger: true,
-        shouldPoll: () => 5000,
```
Comment (tnagorra) [inline src/views/Home/RecentActivity/index.tsx:34]: Any reason why polling was removed?

File: src/views/Leads/Grid/index.js [hunk 8]
```diff
@@ -72,6 +75,23 @@ export default class LeadGrid extends React.Component {
...
+
+                    this.scrollTimeout = window.setTimeout(() => {
+                        this.setState({ showGotoTopButton: e.target.scrollTop > 0 });
+                        console.warn('scroll pos', e.target.scrollTop);
+                    }, 200);
+                };
+
+                c.addEventListener('scroll', this.handleMasonryScroll);
+            }
+        });
```
Comment (tnagorra) [inline src/views/Leads/Grid/index.js]: Please explicitly define the timeout

File: src/views/Leads/Grid/index.js [hunk 9]
```diff
@@ -127,6 +152,20 @@ export default class LeadGrid extends React.Component {
...
+                c.scrollTo({
+                    top: 0,
+                    left: 0,
+                    behavior: 'smooth',
+                });
+            }
+            console.warn('in-loop ref', c);
+        };
+        console.warn('ref', this.masonryRef.current);
+        console.warn('state', this.state.showGotoTopButton);
```
Comment (tnagorra) [inline src/views/Leads/Grid/index.js]: No console.warn meant for debugging

File: src/views/Leads/Grid/index.js [hunk 10]
```diff
@@ -127,6 +152,20 @@ export default class LeadGrid extends React.Component {
...
+            const c = this.masonryRef.current;
+            if (c) {
+                c.scrollTo({
+                    top: 0,
+                    left: 0,
+                    behavior: 'smooth',
+                });
+            }
+            console.warn('in-loop ref', c);
+        };
```
Comment (tnagorra) [inline src/views/Leads/Grid/index.js]: Why are we creating a function on render method

File: src/views/Leads/Grid/index.js [hunk 11]
```diff
@@ -148,6 +187,13 @@ export default class LeadGrid extends React.Component {
                             state={this.itemState}
                             hasMore
                         />
+                        { this.state.showGotoTopButton &&
```
Comment (tnagorra) [inline src/views/Leads/Grid/index.js]: Please de-structure from this.state

File: src/views/Leads/index.js [hunk 12]
```diff
@@ -336,6 +339,26 @@ export default class Leads extends React.PureComponent {
...
+        window.setTimeout(() => {
+            const c = this.tableContainerRef.current;
+
+            if (c) {
+                const sw = c.getElementsByClassName('raw-table-scroll-wrapper')[0];
+
+                this.handleRawTableScroll = (e) => {
+                    window.clearTimeout(this.scrollTimeout);
+
+                    this.scrollTimeout = window.setTimeout(() => {
```
Comment (tnagorra) [inline src/views/Leads/index.js]: Clear this timeout on componentWillUnmount

File: src/views/Leads/index.js [hunk 13]
```diff
@@ -600,6 +658,20 @@ export default class Leads extends React.PureComponent {
...
+                const sw = c.getElementsByClassName('raw-table-scroll-wrapper')[0];
+                if (sw) {
+                    sw.scrollTo({
+                        top: 0,
+                        left: 0,
+                        behavior: 'smooth',
+                    });
+                }
+            }
+        };
```
Comment (tnagorra) [inline src/views/Leads/index.js]: Why create this function on render method

File: src/views/Leads/index.js [hunk 14]
```diff
@@ -653,6 +725,13 @@ export default class Leads extends React.PureComponent {
                                     onPageClick={this.handlePageClick}
                                     onItemsPerPageChange={this.handleLeadsPerPageChange}
                                 />
+                                { this.state.showGotoTopButton && (
```
Comment (tnagorra) [inline src/views/Leads/index.js]: Please de-structure from this.state

**Review by tnagorra — CHANGES_REQUESTED** [review]
Need to clear the timeouts before merge!

File: src/views/Home/RecentActivity/index.tsx (hunk 7 above)
Comment (subinasr) [reply src/views/Home/RecentActivity/index.tsx:34]: We decided to remove it for now. I did mess up the branches though. But, this is okay.

File: src/views/Entries/index.js [hunk 15]
```diff
@@ -525,7 +533,38 @@ export default class Entries extends React.PureComponent {
...
+                const c = this.listContainerRef.current;
+
+                if (c) {
+                    this.handleListScroll = (e) => {
+                        window.clearTimeout(this.scrollTimeout);
+
+                        this.scrollTimeout = window.setTimeout(() => {
+                            this.setState({ gotoTopButtonVisible: e.target.scrollTop > 0 });
+                        }, 200);
+                    };
```
Comment (tnagorra) [inline src/views/Entries/index.js]: Why not define this directly on the class?

File: src/views/Leads/Grid/index.js [hunk 16]
```diff
@@ -72,6 +75,22 @@ export default class LeadGrid extends React.Component {
...
+            const c = this.masonryRef.current;
+
+            if (c) {
+                this.handleMasonryScroll = (e) => {
+                    window.clearTimeout(this.scrollTimeout);
+
+                    this.scrollTimeout = window.setTimeout(() => {
+                        this.setState({ showGotoTopButton: e.target.scrollTop > 0 });
+                    }, 200);
+                };
```
Comment (tnagorra) [inline src/views/Leads/Grid/index.js]: Why not define this directly on the class?

File: src/views/Leads/index.js [hunk 17]
```diff
@@ -336,6 +339,26 @@ export default class Leads extends React.PureComponent {
...
+
+            if (c) {
+                const sw = c.getElementsByClassName('raw-table-scroll-wrapper')[0];
+
+                this.handleRawTableScroll = (e) => {
+                    window.clearTimeout(this.scrollTimeout);
+
+                    this.scrollTimeout = window.setTimeout(() => {
+                        this.setState({ showGotoTopButton: e.target.scrollTop > 0 });
+                    }, 200);
```
Comment (tnagorra) [inline src/views/Leads/index.js]: Why not define this directly on the class?

File: src/views/Leads/index.js [hunk 18]
```diff
@@ -524,6 +554,34 @@ export default class Leads extends React.PureComponent {
...
+                if (c) {
+                    const sw = c.getElementsByClassName('raw-table-scroll-wrapper')[0];
+
+                    this.handleRawTableScroll = (e) => {
+                        window.clearTimeout(this.scrollTimeout);
+
+                        this.scrollTimeout = window.setTimeout(() => {
+                            this.setState({ showGotoTopButton: e.target.scrollTop > 0 });
+                        }, 200);
+                    };
```
Comment (tnagorra) [inline src/views/Leads/index.js]: Why not define this directly on the class?

File: src/views/Leads/index.js [hunk 19]
```diff
@@ -522,14 +551,54 @@ export default class Leads extends React.PureComponent {
...
     }
 
+    handleRawTableScroll = (e) => {
+        window.clearTimeout(this.scrollTimeout);
+
+        this.scrollTimeout = window.setTimeout(() => {
+            this.setState({ showGotoTopButton: e.target.scrollTop > 0 });
+        }, 200);
+    };
+
```
Comment (tnagorra) [inline src/views/Leads/index.js:553]: Looks like this method is duplicated

**Pull request merged**